### PR TITLE
Create/update by NavigationLink identity

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Optimization
- [x] Documentation Update

## Description

Prior to this commit, we were treating all NavigationLink attributes as
unique.  So, were we to change a position of one of the NavigationLinks
during the add_navigation_links rake task, we would have created a new
NavigationLink (and the only one would have remained).

With this commit, we're introducing the concept of the NavigationLink's
surrogate identity, that is to say if we have two NavigationLink objects
with the same `url` and `name` we should consider them the same
NavigationLink.  This allows us to update properties of those
NavigationLinks (via the rake task) without the risk of creating new
entries.

## Related Tickets & Documents

This unblocks PR #16268 which addresses #16076.

## QA Instructions, Screenshots, Recordings

This further helps address some subtle edge cases in our navigation management.

### UI accessibility concerns?

None.

## Added/updated tests?

- [X] Yes

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

**Review by citizen428 — APPROVED** [review]

File: lib/tasks/add_navigation_links.rake [hunk 1]
```diff
@@ -53,56 +53,56 @@ namespace :navigation_links do
...
     task readinglist: :environment do
-      NavigationLink.where(url: "/readinglist").first_or_create(
-        name: "Reading List",
+      NavigationLink.create_or_update_by_identity(
         url: URL.url("readinglist"),
+        name: "Reading List",
         icon: reading_icon,
         display_only_when_signed_in: true,
-        position: 0,
+        position: 1,
```
Comment (joshpuetz) [inline lib/tasks/add_navigation_links.rake:61]: Just curious: did position change from being 0 based to 1 based?

**Review by joshpuetz — APPROVED** [review]

File: lib/tasks/add_navigation_links.rake (hunk 1 above)
Comment (jeremyf) [reply lib/tasks/add_navigation_links.rake:61]: Yes, that's a subtle conversation in some thread I found; namely "Programmers think of staring at 0 but most people won't."  The position shows up in the UI…so here we go.
